projects/sandbox/datagen/datagen/utils/timeslide_waveforms.py

```diff
@@ -0,0 +1,135 @@
+from pathlib import Path
+from textwrap import dedent
+from typing import List, Tuple
+
+import h5py
+import numpy as np
+import torch
+
+from ml4gw.spectral import normalize_psd
+
+
+def calc_shifts_required(
+    segments: List[Tuple[int, int]], Tb: float, shift: float
+):
+    """
+    Based off of the lengths of the segments and the
+    amount of data that will need to be sloughed off
+    the ends due to shifting, calculate how many shifts
+    will be required to achieve Tb seconds worth of background
+
+    Args:
+        segments: A list of tuples of the start and stop times of the segments
+        Tb: The amount of background data to generate
+        shift: The increment to shift the data by
+
+    Returns the number of shifts required to achieve Tb seconds of background
+    """
+
+    livetime = sum([stop - start for start, stop in segments])
+    n_segments = len(segments)
+    shifts_required = 0
+    while True:
+        max_shift = shift * shifts_required
+        total_livetime = (livetime - n_segments * max_shift) * shifts_required
+        if total_livetime < Tb:
+            shifts_required += 1
+            continue
+        break
+
+    return shifts_required
+
+
+def merge_output(datadir: Path):
```

Comment:
Ran some quick tests, first big thing that jumps out at me is that if you can have some way of knowing how many _total_ injections you'll be writing and allocate the memory in the datasets up front, that's about twice as fast as iteratively re-allocating memory in the dataset objects